applications/ehr_query_llm/lmm/rag/ehr/finetune/finetune_embeddings.py

```diff
@@ -0,0 +1,282 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# BE SURE TO HAVE FHIR APP RUNNING BEFORE STARTING THIS SCRIPT
+
+import json
+import random
+import signal
+import sys
+import time
+import urllib
+import uuid
+from threading import Thread
+
+import requests
+import yaml
+import zmq
+from llama_index.finetuning import EmbeddingQAFinetuneDataset, SentenceTransformersFinetuneEngine
+from llama_index.schema import MetadataMode, TextNode
+from tqdm import tqdm
+
+from operators.ehr_query_llm.fhir.ehr_query import FHIRQuery
+from operators.ehr_query_llm.message_handling import MessageReceiver
+
+# Need to run:
+# /opt/nvidia/holoscan/llama.cpp/build/bin/server -m  /workspace/volumes/models/mixtral-slimorca-8x7b.Q5_K_M.gguf --host 0.0.0.0 -ngl 1000 -c 3000
+
+# FHIR global vars
+topic_request = "ehr-request"
+topic_response = "ehr-response"
+busy = False
+resources_to_retrieve = ["Observation", "FamilyMemberHistory", "Condition"]
+ehr_data = None
+
+# Vector db global vars
+embedding_model_cache = "/workspace/volumes/models"
+persist_directory = "/workspace/holohub/applications/ehr_query_llm/lmm/rag/ehr"
+
+# LLM Global Vars
+inference_server_url = "http://0.0.0.0:8080/v1"
+
+
+def get_ehr_data():
+    global ehr_data
+    receiver = MessageReceiver(topic_response, "tcp://localhost:5601")
+    while True:
+        response = receiver.receive_json()
+        if len(response) > 0:
+            for key in response:
+                print(f"Documents retrieved: {len(response[key])}")
+                ehr_data = response
+            return
+        time.sleep(1)
+
+
+def send_request():
+    context = zmq.Context()
+    socket = context.socket(zmq.PUB)
+    socket.bind("tcp://*:5600")
+    # Give ZeroMQ time to bind TODO determine why this is needed
+    time.sleep(1)
+    print("Sending request to FHIR server...")
+    query = FHIRQuery(
+        start_date="1950-01-01", end_date="2023-01-01", resources_to_retrieve=resources_to_retrieve
+    )
+    socket.send_multipart([topic_request.encode("utf-8"), query.to_json().encode("utf-8")])
+
+
+def stop(signum=None, frame=None):
+    print("Stopping...")
+    sys.exit()
+
+
+def update_ehr_dict(flattened_ehr, summary, date_str):
+    """
+    If the summary doesn't exist yet, create a new entry with the provided date. Otherwise,
+    append the date to the existing entry.
+    """
+    if date_str:
+        if summary not in flattened_ehr:
+            flattened_ehr[summary] = [date_str[:10]]
+        else:
+            flattened_ehr[summary].append(date_str[:10])
+    else:
+        if summary not in flattened_ehr:
+            flattened_ehr[summary] = ["N/A"]
+    return flattened_ehr
+
+
+def flatten_ehr(ehr_dict):
```

Comment:
Seems like there is some redundancy in `rag/ehr` flattening script implementations, consider refactoring to one `util` definition? Not a blocker

Reply:
Completely agree.
JIRA task created to track for this:
Need to extract and place the common functions in utility module to avoid duplication and maintenance issues.